### ISSUE TITLE
Proposal: Use 2 threads per CPU core by default

### ISSUE THREAD

**Issue by HadrienG2** [LLNL/spack#26242]
### Summary

I believe that on x86_64 and other 2-way SMT architectures, spack should default to using all available CPU resources, as if called with the `-j $(nproc)` CLI parameter. Highly threaded CPU architectures like POWER, however might benefit from a cap on the amount of threads per core that is used. Benchmarking on such CPUs would be useful to clarify where the default cap should be set, but from x86 experience, it is probably safe to start by setting it to 2.

This proposal is directly challenging the point of view, previously stated in issue #11072 and implemented in PR #11373 that absolute build concurrency limits are a good thing. Some discussion with @jjwilke @mwkrentel and @citibeth would therefore be desirable before any decision is taken.

### Rationale

Unless I have missed something, the motivation behind spack's default cap of 16 build jobs, which is configurable via the `config:build_jobs` setting, is not stated clearly in the documentation.

To understand where this comes from, I therefore had to dig through the issue tracker to finally find issue #11072, which motivated the introduction of the 16-cpu cap through PR #11373.

From my reading of this issue, two different rationales were invoked to justify this decision:
1. Some C++ projects cannot build with all system CPUs because their build processes use up too much memory, leading to exhaustion of available system RAM
2. Building with lots of processes on highly threaded CPUs like POWER chips yields diminishing returns anyway.

---

With all due respect to @jjwilke, who originally formulated it, I think that the first rationale is misguided, in the sense that it blames on a build system a problem that actually originates from the software that is being built, and should therefore lie on the shoulder of said software's developers.

<details>
<summary>Here's why.</summary>

If one thinks for a second about the kind of work that a compiler does (translating and transforming abstract representations of the compilation unit that is being processed) and then ponders the volume of information that can be stored in a gigabyte of memory, it becomes readily apparent that when a compiler starts consuming multiple gigabytes of memory, it can only mean one of a few things:

1. The compiler exhibits a major bug (e.g. memory leak) that must be tracked down and fixed.
2. The compiler makes extraordinarily inefficient use of system resources and should be optimized.
3. The compilation unit that is being processed is extremely large, and its contents should be investigated.

From my experience, and that of other C++ developers fighting build bloat whose testimonies I've found across the Web, the third explanation is usually the right one. If one applies suitable tooling, such as clang's `-ftime-trace` or Templight, to a suspciously costly compilation units, one will almost always find dubious coding practices all across the codebase, which only end up showing in a specific compilation unit more than the others by means of accumulation. These include:

- Undiscriminated use of inline function declarations in headers, in spite of the fact that for most functions this will result in lots of compilation work duplication with no significant associated runtime performance benefit.
- Templated code, which must be inlined with most compilers, is written without any due attention paid to the fact that one copy of the code will be processed by the compiler for each template instantiation. As soon as this consideration is properly taken into account by developers, steps can be taken to reduce the associated code redundancy, revolving around the general idea of extracting the type-generic part of the templated code into a separate function (that may then be moved to a separate compilation unit if appropriate).
- Templates are overused for use cases where they are an unnecessary evil and run-time polymorphism would yield more ergonomic APIs, reduced compilation cost, and comparable run-time performance.
- Evil metaprogramming constructs are seemingly purpose-built to make compile-time code combinatorics explode without providing a matching benefit in exchange (my personal pet peeve here would be Eigen's use of expression templates, which actually provide little run-time performance benefit to match the compile-time combinatorial explosion that they introduce, in the small-matrix use cases that people most often use Eigen for).

From this perspective, reducing build system concurrency in an attempt to avoid crashes caused by excessive compiler memory consumption is more often than not curing the symptom, rather than the disease. It is also a short-term fix in the sense that if the problem is not fixed at the source, which is lack of developer education and attention regarding build-time vs run-time performance tradeoffs, it will only get worse over time until it reaches the point where it becomes unbearable and real solutions need to be explored.

This point will usually be reached as soon as the code needs to be built on new machines whose resources are much more constrained than anything that was previously used. For example, the project may want to use public Github CI VMs (which are equipped with 2 cores and 7 GB of RAM), or a new developer/user may join and desire to build the project using a laptop whose specs are much weaker than what the other developers are used to.

And from painful personal experience on a project where the largest compilation units individually reached 7 GB max-RSS and multi-minutes build time at the time where I had to intervene, it would be much better for everyone if the problem were acknowledged and tackled early on, rather than at the last moment where some people simply cannot work anymore until it is fixed.
</details>

---

The second rationale, on the other hand, I can agree with.

Highly threaded CPU architectures are really meant to shine on IO-bound work, like typical web servers, where CPU execution resources are only lightly used and context switching performance is the name on the game. But from my experience, on typical hardware compilation is not quite I/O bound. It's mostly CPU-bound, it just happens that due to the way compilers are designed, the CPU is not used super-efficiently so there's room for another thread to steal a couple of cycles that would otherwise go to waste.

Therefore, I think it makes sense to set a cap on the amount of **threads per physical core** that spack uses. The limit would need to be tuned based on some benchmarking on highly-threaded CPUs like POWER, but it could be initially set to 2 since it is well-known that on x86, hyperthreading leads to near-perfect scaling for compilation workloads, which leads to an expectation that SMT2 should work as well on POWER.

---

Now, beyond opposing other people's rationale, I must also state mine, so here goes:

- Using all logical CPUs by default is the norm in all modern software construction tools (scons, ninja, cargo, bazel...). This creates a user expectation, which is violated by spack. As a result, spack's ergonomics suffer from this feature.
- To make matters worse, spack's build job cap logic is not immediately obvious to users, as it is not put front and center in the documentation, and the associated behavior is pretty much invisible on the smaller personal machines where users are likely to start experimenting with spack. When these users later move to a beefier build node, they may fail to cross-check that spack uses all their CPU cores (and why would they ?), which will lead to fine compute resources going to waste and frustrated users wondering what's going on.

### Description

If this proposal is accepted, the behavior of `config:build_jobs` would be reverted to where it was before PR #11373, i.e. no user-configured limit means no job concurrency limit at all.

As an alternative, a second configurable, which could perhaps be named `config:build_jobs_per_core`, would be introduced. This setting would default to 2 as a starting point, and might be tuned up later on if further benchmarking reveals that SMT-heavy compilation workloads scale better than expected here (and informally reported on #11072).

Spack would honor this setting by setting a build jobs cap which is the maximum of `build_jobs` (if set) and `build_jobs_per_core` times the number of physical CPU cores.

### Additional information

This problem exists on the current develop version of spack (`0.16.3-4390-b056e871f8`). It was introduced by PR #11373, in an attempt to fix issue #11072. While the topic of this issue is, in a sense, very close to that of #11072, I feel that so much time has passed and the point of view presented here is so different that necroposting by reopening #11072 might not be desirable.

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by haampie** [comment]
I tend to agree with this for the most part, in particular with making Spack's handling of build jobs less surprising compared to established build systems. Are there any known build systems that actually set a default cap on the number of jobs?

Couple notes:

1. I think another reason to do have a configurable cap is to avoid "abuse of the login node" in HPC centers. But well, maybe centers should stop living in the past... In any case, some sysadmins would be disappointed if there was no absolute cap on the number of jobs, and they may as well set it to max 2 if they are fed up with compilation jobs on shared resources :p.
2. `spack install -j ...` overrides the number of jobs even if they exceed the cap in the config (https://github.com/spack/spack/pull/22360)
3. You have to consider that an increasingly popular compiler in the Spack community is not particularly well known for its efficiency (and there may be more closed source compilers that have incredible performance claims that are largely due to inlining everything): https://github.com/ROCm-Developer-Tools/HIP/issues/2285. So, these compilers exist and we have to live with them.
4. Many recent HPC centers now come with (dual socket) AMD EPYC cpus and a lot of memory, say 256 threads per node. This was not yet the case in the time when the max 16 jobs limit was set. Maybe this could change our opinion about the number 16. At the same time I'm not sure if going for 256 jobs by default is best if config:build_jobs_per_core:2 was the default.
5. When using `slurm`, `docker` or `taskset`, you create processes that have limited access to cores through cgroups. Here you have to look at what number of cores/threads is available to your process, not the number of physical cores. Spack currently already does this.

**Comment by mwkrentel** [comment]
I'll just remind everyone that with the previous defaults, I tried
running `spack install hpctoolkit` on one of our local clusters and it
wedged the machine hard enough that the admins had to reboot it
because spack ran `make -j 192`.

It's Ok to provide the knobs/config to shoot yourself in the foot or
piss of the admins or get your account suspended.  But it can't be the
default, you have to explicitly ask for it.

If it were me, I'd set the defaults for a high number of threads (>64)
at something conservative, not hobbled, just conservative.  Maybe 1/3
of available threads so as not to hog or overload the machine.
I'd also obey `build_jobs` or `-j`, but they have to be explicitly set.
Just my humble opinion.

**Comment by HadrienG2** [comment]
I've been thinking about these some more, and here are some thoughts about some of the concerns raised by @haampie:

> I think another reason to do have a configurable cap is to avoid "abuse of the login node" in HPC centers. But well, maybe centers should stop living in the past... In any case, some sysadmins would be disappointed if there was no absolute cap on the number of jobs, and they may as well set it to max 2 if they are fed up with compilation jobs on shared resources :p.

Yes, this is why I kept the option of letting sysadmins manually tune a `build_jobs` cap in the config file. Though if I were them, I would fix the "abuse of login node" problem in lower-level ways that apply to every app, not just Spack:

- If the problem is that by running CPU-intensive jobs on the login nodes, users make said nodes unacceptably slower for everyone else, then an app-generic solution would be to address this at the Linux scheduler level. One way is to downgrade the priority / increase the niceness of long-lived CPU-intensive processes. I don't remember from memory if CFS has a tunable to do this on its own, but if it's doesn't, this doesn't sound too difficult to do from the same sort of system monitoring daemons that are sometimes used to auto-ban "abusive" users.
- If the sysadmin just has a general beef against users consuming lots of CPU on login nodes, and wants to aggressively hint them towards doing their builds on compute nodes even if the other CPUs are currently sitting idle, a more aggressive option would be to use cgroups to set a max CPU budget on each individual user session. One relatively ergonomic way to do this is to apply [systemd resource control limits](https://www.freedesktop.org/software/systemd/man/systemd.resource-control.html) to each individual user slice (`user-xyz.slice`), which itself can be done by putting a suitable configuration file into `/etc/systemd/system/user-.slice.d/*.conf`.

> `spack install -j ...` overrides the number of jobs even if they exceed the cap in the config (#22360)

Good point. I don't think this interacts with the present proposal though, since it operates at the "lower-priority" configuration file layer.

> You have to consider that an increasingly popular compiler in the Spack community is not particularly well known for its efficiency (and there may be more closed source compilers that have incredible performance claims that are largely due to inlining everything): ROCm-Developer-Tools/HIP#2285. So, these compilers exist and we have to live with them.

As long as such compilers remain relatively obscure, and promise to fix such bugs in the longer run (as in the ticket that you linked to), I would argue that Spack itself should not handle this, and rather leave workarounds for those sort of specialized compiler bugs up to site-specific configuration. This, again, highlights the importance of having a `build_jobs` tunable, which I am not opposing per se as long as it's off by default.

> Many recent HPC centers now come with (dual socket) AMD EPYC cpus and a lot of memory, say 256 threads per node. This was not yet the case in the time when the max 16 jobs limit was set. Maybe this could change our opinion about the number 16. At the same time I'm not sure if going for 256 jobs by default is best if config:build_jobs_per_core:2 was the default.

Indeed, with that many physical build cores, we are likely to hit the scalability limit of the package build itself (there is just not that many compilation units that can be built in parallel in many software).

I suspect that in such cases, it would be beneficial to not just leave all the parallelism up to each individual package build, but also leverage the other form of build parallelism that Spack supports, namely building multiple packages concurrently. Taking inspiration from MPI+OpenMP codes, a basic static partitioning scheme that could be used here would be to build one package per NUMA node and use cores within the NUMA node for running the build in parallel.

But I also suspect that even with such a relatively simple partitioning scheme (which is pretty much guaranteed to behave badly in typical builds that are occasionally bottlenecked by a single big package, a situation that requires a more dynamic partitioning algorithm for optimal CPU efficiency), this would still make the proposal a lot harder to implement, especially when it comes to keeping stdio logs intelligible...

Thoughts on this?

> When using slurm, docker or taskset, you create processes that have limited access to cores through cgroups. Here you have to look at what number of cores/threads is available to your process, not the number of physical cores. Spack currently already does this.

Yes, I thought about this and also found out that Spack can already query the taskset to figure out how many usable CPU cores it has at its disposal. Don't know if Spack has access to full NUMA domain topology info as well, however, given that this would be needed for the above proposal. Do you know?

---

@mwkrentel If I remember correctly your original report, that build was on an IBM POWER processor. Those really have a huge lot of threads per CPU core (up to 8 from memory), and I suspect that even if the build itself had enough concurrency, using every CPU thread would not be efficient at the hardware level, because the CPU execution backend would likely become fully saturated with only a subset of these threads on.

This is why I am proposing to set a cap on the number of threads per core that is used, which would from my understanding resolve your original problem (assuming RAM/core is reasonably high, in this configuration RAM/thread does not ned to be).

Do you think this would be enough?

**Comment by goxberry** [comment]
I disagree with this proposal simply because I think the speed versus reliability tradeoff is unfavorable; I think #11072 (which I assume is what was really meant when @HadrienG2 linked to #1072) illustrates this point.

The counterarguments I've seen so far to objections or potential objections to this proposal seem like they reduce to "any system crashing/wedging/slowdowns resulting from this change should be addressed through means external to the project, because those means are better places to address those undesirable behaviors." If a user needs to install a package right away -- which is probably why they're using Spack in the first place -- it's not great user experience to crash their system and then tell them they should contact some other person who may or may not fix that issue. If my experience with my co-workers using Spack is any indication, those users are likely to blame Spack for the crash, whether or not that blame is deserved, and they are unlikely to find minimizing build times sufficiently compelling either.

**Comment by HadrienG2** [comment]
@goxberry Thanks for the correction #1072 -> #11072, looks like a typo that got propagated everywhere. Fixed it in the OP.

---

I think you hit the nail in the head when you say that this is a speed vs reliability tradeoff. The problem is that as with any quantitative tradeoff, as long as people's opinions are our only metric, we will have issues because these opinions will fall at different points across that tradeoff.

At one extreme, you will find people who want maximal safety by defaulting to 1 build job. But that yields extremely slow and inefficient builds on modern hardware. And it is not a perfect safeguard against OOM crashes either: a pathological C++ project that I work on wouldn't even build at -j1 on a system with 7GB of RAM a while ago, which prevented use of Github CI on it before we optimized the build's resource consumption.

At the other extreme, you will find people like me who think that idle build farm CPUs are a waste of hardware and developer time. From this perspective, if your build crashes from OOM at the concurrency level that is optimal from a CPU efficiency point of view, then this must be fixed by either installing more RAM or optimizing the software build's resource usage. The build system is not in a good position to make a decision here, because normally it knows nothing about the RAM vs CPU tradeoff, and therefore cannot tell how an amount of build processes will translate into a RAM consumption figure (but see below).

The question then is, how do you arbitrate across this tradeoff to pick a wise default that will yield a good speed/reliability tradeoff most of the time?

---

In my opinion, quantitative tradeoffs are best resolved with empirical data. For example, in the aforementioned "buy extra RAM" vs "optimize software build consumption" tradeoff, I picked a limit through the following process...

- The overwhelming majority of computers that I see used to build software around me (WLCG, HPC centers, developer personal machines...) are x86 machines configured with >=2GB of RAM per CPU thread.
- Developer machines with less than 4 CPU threads (either 2C+HT or 4C w/o HT) are getting very rare, and are therefore probably not worth optimizing for anymore, so let's take 4 threads as our baseline.
- This means that you can assume you have at least 8GB of RAM at hand, of which OS and background apps will take 1-2 GB even if user is the careful to run nothing but the build + their desktop environment.
- From this data, I conclude that software builds which use over (8-2)/4 = 1.5 GB per core are pathological and must be optimized, on the grounds that upgrading the RAM of every build machine out there or excluding lower-end dev machines is not a reasonable option, and wasting CPU resources in builds means wasting people's time, which is provably the single most expensive resource that a software project has to manage.

---

From this empirical perspective, given that we can't expect Spack packages to provide memory consumption figures, let alone reliable ones, I believe the wisest starting point if you wanted safer spack defaults than what I am proposing would be to pick said default based on what is most likely to work on the build host.

On smaller personal developer machines, fully parallel builds are the norm. Therefore, we can expect that the majority of software has been tested (if involuntarily) to build correctly with 1.5GB of RAM per CPU thread. If you wanted a safer spack default with minimal code cost, one simple approach would therefore be to query the amount of available RAM at the time where spack was started, and tune the number of build processes according to this criterion in addition to the number of available CPUs.

Personally, I would also print a warning if this ends up selecting a sub-optimal process count from a CPU efficiency PoV. But most importantly, this would not yet address the issue of badly behaved software builds.

To handle those gracefully would actually be possible, but require a lot more code complexity on the spack side. The idea would be to detect when the aforementioned expectation of sane RAM usage is violated, by monitoring the RAM usage of build jobs and overall system memory pressure:

- Keep track of the memory usage of build processes over time, use this to evaluate the max RAM/process for this particular build.
- If the system memory usage goes above a high water mark (say, max-0.5GB), then...
    1. Stop the build.
    2. Print a warning message that explains the problem that was encountered (either insufficient RAM/core due to excessive background system activity, or pathological software build, depending on the 1,5GB/thread criterion).
    3. Restart the build with a number of processes that should be safe according to the previous RAM/process measurement.
- If the build fails again with this auto-tuned number of process, then either repeat the cycle with new system resource usage estimate, or just go for plain `-j1` straight away.
- If even `-j1` doesn't work, abort the build altogether, and ask the user to turn off background apps or report a bug to the software that was being built depending on whether the problem was baseline RAM usage or the build's RAM/thread usage.

All of this would concretely improve the safety of spack builds against OOM on any computer, which would be a safety improvement over the current `build_jobs: 16` default that actually does not help at all on smaller computers (badly behaved C++ code will very easily OOM machines with much less than 16 hardware threads).

**Comment by goxberry** [comment]
tl;dr: The current defaults for 16 or fewer logical cores are fine. I favor @mwkrentel's [suggestion](https://github.com/spack/spack/issues/26242#issuecomment-927022954) for larger core-count systems.

-----------------

> ...a pathological C++ project that I work on wouldn't even build at -j1 on a system with 7GB of RAM a while ago...

In general, fixing a problem upstream is always preferable to kludging around it in Spack. If RAM requirements to builds are sufficiently large, even in the `make -j1` case, that package should be considered buggy/broken.

> At the other extreme, you will find people like me who think that idle build farm CPUs are a waste of hardware and developer time.

So is bikeshedding. In my opinion, attempting to maximize performance via defaults is inappropriate: defaults should aim for "pretty good" settings. Users still have the ability to use Spack's configuration files and command-line options to tune more aggressively for performance.

> The idea would be to detect when the aforementioned expectation of sane RAM usage is violated, by monitoring the RAM usage of build jobs and overall system memory pressure: ...

Here's a simpler counterproposal that's partially RAM-based, but would require vending in the [`psutil` package](https://github.com/giampaolo/psutil) (and probably some sanity-checking/tuning, modulo bugs):

```python
if cpus_available() <= 16:
    config_defaults['config:build_jobs'] = min(cpus_available(), 16)
else:
    bytes_per_gigabyte = 1e9
    ram_per_core_limit_in_gigabytes = 3
    ram_per_core_limit_in_bytes = ram_per_core_limit_in_gigabytes * bytes_per_gigabyte
    max_fraction_cores_used = 0.5
    possible_default_job_limits = [
        16,
        math.trunc(psutil.virtual_memory().total / ram_per_core_limit_in_bytes),
        math.trunc(cpus_available() * max_fraction_cores_used)
    ]
    possible_default_job_limits.sort()
    median_index = 1
    config_defaults['config:build_jobs'] = possible_default_job_limits[median_index]
```

Note that [`psutil.virtual_memory().total`](https://psutil.readthedocs.io/en/latest/index.html?highlight=memory#psutil.virtual_memory) returns total physical memory (in bytes). I still like @mwkrentel's proposal better -- it's simpler -- but if the main objection is really "let's guard against OOM and use more cores", maybe including a RAM-based job limit criterion addresses that objection more directly.

**Comment by HadrienG2** [comment]
tl;dr: How about `build_jobs = min(2 * physical_cpu_cores, physical RAM / 2GB)`?

---

@goxberry Sorry, I do not understand what your latest proposal is trying to achieve.

Earlier, you said that your main issue with my initial proposal was that it put unknowing users at the risk of crashing their system, which would cause bad user experience and ultimately bad reputation for Spack.

> If a user needs to install a package right away -- which is probably why they're using Spack in the first place -- it's not great user experience to crash their system and then tell them they should contact some other person who may or may not fix that issue. If my experience with my co-workers using Spack is any indication, those users are likely to blame Spack for the crash, whether or not that blame is deserved, and they are unlikely to find minimizing build times sufficiently compelling either.

Accepting this argument, I tried to adress the question of how spack's default could do their best to prevent OOM crashes, without overly compromizing the default build performance for the sake of a couple of pathological builds (with which I have personal experience), and stated the following facts:

- By and large, today's standard for build machines is x86 with **2 GB/thread** of RAM.
- Most personal developer machines (on which package builds are tested daily, and which many Spack users will use), don't have many cores. The baseline for laptops in the wild is **4 threads** in either 2C+HT or 4C configuration.
- These **smaller machines are at the highest risk of failing from OOM** when running spack, because the OS + graphical environment consumes a significant fraction of the available RAM. Hence I would recommend not using more than **1.5 GB RAM / thread** if you want to avoid OOM on them.

You did not challenge those facts, but I do not see how your proposal takes them into account either:

- You keep the current **"max concurrency until 16 build jobs"** magic. But **this policy maximizes the risk of build OOM on smaller machines**. Which are, as outlined above, those at the highest risk of OOMing to begin with.
- For larger machines, you use a **3 GB/process RAM budget**. This seems strangely pessimistic, given the numbers above. Software is built daily on machines with with 2 GB/thread, and that's an amount of RAM which is only available asymptotically when the system has so many cores and is so carefully tuned that system/daemon RAM consumption is negligible. Most software is actually tested to build on systems with a 1.5 GB/thread RAM budget.
- You also set a hard limit on **not using more than half of the system's threads**. I do not understand what this is trying to achieve. It will under-utilize CPUs in the common case of x86 machines, where using all the threads is a good idea. And in the more exotic case of highly-threaded POWER CPUs, which motivated having a more complex policy than "use all the threads" to begin with, it may have the reverse effect of over-subscribing the system.
    * I thought maybe your goal would be to leave enough CPU capacity for good interactive performance without manual sysadmin tuning of spack process priority. But that seems unnecessarily distrustful of the Linux scheduler to me. I've done countless Zoom conferences with Spack using up all the CPUs in the background, and even in this very latency-sensitive task the performance remained tolerable, if not great. In those rare cases where I really needed to keep the system 100% snappy while a build is running, setting aside 1-2 CPU threads for interactive processes was enough.
- In the end, those three limits are combined into a single decision using a non-linear formula. The more we do this sort of things, the more the policy becomes impenetrable for users. For this reason, I initially wished we could get away with a default policy that has a single parameter (number of processes per physical core), like the current "max 16 jobs" policy. I can accept that, for OOM safety reasons, we might want to have a default policy with two parameters (e.g. number of processes per physical core and minimum RAM budget per process). But as you point out yourself, **we should hopefully not need a level of complexity that calls for three tunable parameters in our default**.

And on this simplicity front, you say another thing that I couldn't agree with more:

> > At the other extreme, you will find people like me who think that idle build farm CPUs are a waste of hardware and developer time.
>
> So is bikeshedding. In my opinion, attempting to maximize performance via defaults is inappropriate: defaults should aim for "pretty good" settings. Users still have the ability to use Spack's configuration files and command-line options to tune more aggressively for performance.

So here's a proposal which is I believe is as simple as possible while acknowledging all concerns expressed so far fairly: `build_jobs = min(2 * physical_cpu_cores, physical RAM / 2GB)`.

This proposal acknowledges that developers normally test their builds on x86 systems with 2-way SMT and a RAM budget of 2GB/thread by not attempting to use more than 2-way SMT and reserving a RAM budget of 2GB/process. Hence...

- It should be OOM-safe on non-pathological builds that work on typical dev machines. Pathological builds can tune down `build_jobs` as a workaround, but @goxberry agreed with me that overly optimizing for them on the Spack side is unwise.
- It should be optimal on most x86 systems, with the possible exception of the big EPYC nodes mentioned by @haampie , which will run out of intra-package build concurrency (in this case, cores simply go unused by make/ninja/spack, which is not great but not that bad either). This can later be improved by making Spack more clever about parallel package builds.
- It should be safe and not too inefficient on POWER systems, as it acknowledges that software builds are only tested on 2-way SMT x86 systems and does not attempt to use more than 2-way SMT.

From the simplicity point of view, I also agree that making the 2 threads/cores and 2GB/process tunable via Spack configuration is overkill. Just use them in the default, and forget about them when the user overrides `build_jobs` manually.

**Comment by mwkrentel** [comment]
Geez, who knew it was so difficult to come up with a simple default.

I think it's very risky to argue, "most modern systems have ..., so
let's push the limit to ...."  There is old hardware, new hardware,
single-user systems, sytems with 20 other users running -j 32.  A few
years ago, TACC disabled SMT on their systems, so "2 times number of
cores" would not be such a good idea there.

The point is that it's too hard to predict how far you can push the
limit on a given system, especially if you don't know if you're the
only user or not.  So don't try.

There are two extremes that you want to avoid.

1. You don't want to run at -j 1.  So, on systems with a "low" number
of cores/threads (up to 8 or 16), use all available threads.  This
gets you at least 75% of available parallelism.

2. You don't want to overwhelm the system, especially if you're not
the only user.  So, on systems with a "high" number of cores/threads
(32 or 64 or more), use maybe 50% of cores with a cap of maybe 32.

Beyond that, if a higher number is appropriate, then let the user
specify it him/herself.

I tried a simple experiment on a local power9 system with 40 cores,
160 threads and 256 Gig RAM.  The time to build hpctoolkit at -j 16
was 20 min, 20 sec.  At -j 64 it dropped to 18:41.  So, 4X the number
of threads for a whopping 8% faster.

So, I think it's more important to avoid pushing the limits of the
system than it is to squeeze out that last 5%.  Just the "simple"
default.  The user can always crank it up or down as appropriate.

**Comment by goxberry** [comment]
> Sorry, I do not understand what your latest proposal is trying to achieve.

@HadrienG2, the developers I support primarily use HPC nodes and are not sysadmins (nor am I, for that matter), so if they bring down a node, they get scolded by sysadmins and it becomes my problem to debug. I am optimizing for a different case than you are -- namely, avoiding trouble from my colleagues -- which is the root of our disagreement.

>  Hence I would recommend not using more than 1.5 GB RAM / thread if you want to avoid OOM on them.

**I don't particularly care what the default job limit is, so long as it's conservative enough that it doesn't crash an HPC node, even under less than ideal operating conditions, and it's fast enough.** You've allowed seemingly no margin for error on systems that are sometimes finicky to work with, and your rationale for this choice is to cite statistics don't obviously seem to apply to HPC (e.g., Gentoo's similar advice re: memory per thread). The whole point of #11072 involved wedging an HPC node, and belaboring statistics re: developer machines seems to miss that point entirely. To wit: vendor compilers are not "obscure" for HPC, and speculating about them as such is part of the reason you're getting this pushback.

> You did not challenge those facts, but I do not see how your proposal takes them into account either:

I'm not challenging the facts -- I'm challenging the assumptions. I don't even care about the specifics other than having some margin of safety and a policy that is no more conservative than the status quo.

> But as you point out yourself, we should hopefully not need a level of complexity that calls for three tunable parameters in our default.

I fail to see how your suggestion is linear and involves fewer than three parameters, given that your "minimum threshold" is either 0 or 1 instead of 16; you merely picked different values. In light of your earlier discussions of detecting memory pressure and NUMA topology, complexity is a strange criticism to level here.

> And in the more exotic case of highly-threaded POWER CPUs, which motivated having a more complex policy than "use all the threads" to begin with, it may have the reverse effect of over-subscribing the system

The POWER machines I've worked with typically have 1 GB RAM/thread or less. On LLNL's Lassen, for instance, a node has 44 POWER9 CPUs with 8-way SMT (352 logical CPUs) and 256 GB RAM -- around 0.73 GB RAM/logical-CPU. The Spack `cpus_available` method returns the number of logical cores available, so my proposal would use 85 jobs, while yours would use 88 jobs. I fail to see how my proposal would oversubscribe this system while yours would not; note also that all proposals except yours also account for varying levels of SMT in an architecture-consistent way.

> In my opinion, quantitative tradeoffs are best resolved with empirical data.

I agree, so I'll ask for it: are there compelling examples where running `make -j64` or `make -j128` instead of `make -j16` on an HPC node with sufficient resources matters often for the average user?
 
Fixing problems "upstream" is, of course, better than kludging around them with conservative settings. Nevertheless, these problems do happen, and upstream fixes do not necessarily land in a timely fashion, if at all. For these reasons, I don't understand the willingness to downplay or even dismiss these real obstacles, the potential support burden they incur, and any sort of less aggressive compromise, especially when other maintainers are likely to have to deal with the consequences.

I doubt my perspective will change with further posts on this thread, so this reply will be my last of this issue. I agree with @mwkrentel & @haampie, and leave continuation of the debate to others.

**Comment by HadrienG2** [comment]
Oh well, if...

- Everyone is quite obviously getting fed up with this discussion.
- It seems no simple policy will please everyone.
- We just don't agree enough on the tradeoffs & priorities involved to pick a non-simple policy.

Then it's better to just call it a day.

Just to fix what I perceive as a misunderstanding: it's not as if I care only about optimal build perf, or don't care at all about OOM safety, or whatever. As the author of this proposal, I felt it was my duty to consider each goal brought by other participants in turn, see what kind of policy it would take to fully optimize for it, then attempted to find a balance between these various optimal policies + the extra goal of keeping things simple.

This is why I started with complex solutions, then tried to find simpler ones that still get the basics right.
